### PR TITLE
Add image restoration losses: Charbonnier, PSNR, TotalVariation, EdgeAwareSmoothness, MSSSIM

### LINKED ISSUE #23126: [Feature request] A Restoration and Perceptual loss suite
## Summary

Adding a small suite of high-level image restoration losses to `keras.losses`:

- `keras.losses.Charbonnier`
- `keras.losses.MSSSIM`
- `keras.losses.PSNR`
- `keras.losses.TotalVariation`
- `keras.losses.EdgeAwareSmoothness`
- And an optional: a KerasCV-backed `PerceptualLoss` / LPIPS-like wrapper when a pretrained feature extractor is available

## Motivation

Image restoration workflows commonly combine pixel, perceptual, smoothness, and image-quality losses. Today, Keras users often hand-roll these losses even though the formulas are stable and easy to test.

This creates friction for tasks such as:

- denoising
- deblurring
- super-resolution
- demosaicing / neural ISP
- inverse rendering
- depth or disparity smoothness
- plug-and-play restoration

Recent restoration and imaging work continues to rely on these loss families, including:

- SP3: Spherical Priors for Plug-and-Play Restoration
- DDTNet
- NTIRE 2026 Image Denoising Challenge report
- Geodesic Flow Matching on a Riemannian Degradation Manifold
- The Need for Neural ISP in the Small-Pixel Era

Other libraries already expose similar functionality. For example, Kornia includes restoration-oriented losses such as SSIM3D, MS-SSIM + L1, TotalVariation, PSNRLoss, InverseDepthSmoothnessLoss, and CharbonnierLoss. Keras should offer the same kind of ergonomic entry point while preserving Keras' backend-agnostic `keras.ops` design.

## Proposed API

```python
keras.losses.Charbonnier(
    epsilon=1e-3,
    reduction="sum_over_batch_size",
    name="charbonnier",
)

keras.losses.MSSSIM(
    max_val=1.0,
    alpha=0.84,
    l1_weight=0.0,
    axis=(-3, -2, -1),
    reduction="sum_over_batch_size",
    name="msssim",
)

keras.losses.PSNR(
    max_val=1.0,
    axis=(-3, -2, -1),
    reduction="sum_over_batch_size",
    name="psnr",
)

keras.losses.TotalVariation(
    axis=(-3, -2),
    reduction="sum_over_batch_size",
    name="total_variation",
)

keras.losses.EdgeAwareSmoothness(
    axis=(-3, -2),
    reduction="sum_over_batch_size",
    name="edge_aware_smoothness",
)
```

Optional KerasCV integration:

```python
keras.losses.PerceptualLoss(
    feature_extractor,
    layers=None,
    weights=None,
    reduction="sum_over_batch_size",
    name="perceptual_loss",
)
```

This could live behind KerasCV or only be enabled when a pretrained feature extractor is supplied, avoiding a hard dependency in core Keras.

### Charbonnier

```python
sqrt((y_pred - y_true) ** 2 + epsilon ** 2)
```

Charbonnier is a smooth L1-like loss commonly used in restoration because it is less sensitive to outliers than MSE while remaining differentiable around zero.

### MS-SSIM / MS-SSIM + L1

```python
loss = alpha * L_MSSSIM + (1 - alpha) * G_alpha * L1
```

`MSSSIM` should support a plain MS-SSIM loss and, optionally, the common MS-SSIM + L1 variant through `l1_weight` or equivalent weighting.

### PSNR

```python
psnr = 20 * log10(max_val) - 10 * log10(mse)
loss = -psnr
```

PSNR is usually reported as a metric, but a loss wrapper is useful for consistency with other restoration libraries and for simple optimization baselines.

### Total Variation

```python
sum(abs(x[:, 1:] - x[:, :-1])) + sum(abs(x[:, :, 1:] - x[:, :, :-1]))
```

The implementation should generalize across supported image data formats through `axis`.

### Edge-Aware Smoothness

A common form is:

```python
abs(dx_pred) * exp(-abs(dx_image)) + abs(dy_pred) * exp(-abs(dy_image))
```

This is widely used for depth, disparity, optical flow, and other spatial prediction tasks where smoothness should be discouraged across strong image edges.

## Testing

- zero loss for identical inputs where appropriate
- finite outputs for small epsilon / zero-difference cases
- correct sign and monotonicity for PSNR loss
- shape/reduction behavior matches existing Keras losses
- channel-last and channel-first coverage where supported
- dtype coverage for float inputs
- backend-agnostic tests using `keras.ops`

## Why??

These losses are useful across image restoration tasks. They are standard building blocks that users repeatedly recreate.

Adding them to Keras would reduce duplicated user code
This is a low-risk addition with a clear user benefit and a small implementation surface.

If the feature seems useful, I can start with building. Thank you

### PR DESCRIPTION
**Pull request by pctablet505**
## Summary

This PR implements the core image restoration losses requested in #23126:

- `keras.losses.Charbonnier`
- `keras.losses.PSNR`
- `keras.losses.TotalVariation`
- `keras.losses.EdgeAwareSmoothness`
- `keras.losses.MSSSIM`

Each loss is exposed both as a standalone function (`keras.losses.charbonnier`, etc.) and as a `Loss` class that supports Keras reduction, dtype, serialization, and the `compile()` API.

## What is implemented

- **Charbonnier** — smooth L1-like loss: `mean(sqrt(diff^2 + epsilon^2))`.
- **PSNR** — negative PSNR computed as a loss, with per-image MSE reduction over the image axes.
- **TotalVariation** — sum of absolute spatial differences, regularizing predicted images.
- **EdgeAwareSmoothness** — smoothness of predictions weighted by `exp(-|gradient(y_true)|)`, preserving ground-truth edges.
- **MSSSIM** — proper multi-scale SSIM loss using per-channel Gaussian-filtered luminance and contrast-structure components, plus an optional MS-SSIM + L1 combination controlled by `alpha` and `l1_weight`.

All five classes are registered in `keras.src.losses.__init__.py` and the generated public API (`keras/api/losses/__init__.py`, `keras/api/_tf_keras/keras/losses/__init__.py`).

## Tests

Added comprehensive real-value tests in `keras/src/losses/losses_test.py`:

- `CharbonnierTest`
- `PSNRTest`
- `TotalVariationTest`
- `EdgeAwareSmoothnessTest`
- `MSSSIMTest`

Each covers class serialization (`test_config`), correctness against hand-computed values, reduction modes, dtype arguments, function forms, and backend-agnostic behavior (verified on NumPy, TensorFlow, and JAX backends).

## Verification

- `pre-commit run --all-files` passes (api_gen, ruff, ruff-format).
- `pytest keras/src/losses/losses_test.py` passes on NumPy, TensorFlow, and JAX backends.

## Notes / not implemented

- `PerceptualLoss` / LPIPS was listed as optional in the issue and is left for a follow-up because it depends on a pretrained feature extractor / KerasCV integration.

Closes #23126 (partial; perceptual loss deferred).

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

**Comment by hertschuh** [comment]
Per discussion in issue, we will not be adding these to core keras.
